marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/entities/entity-search-lib.sjs

```diff
@@ -141,6 +153,81 @@ function getPropertyValues(currentProperty, entityInstance) {
   return resultObject;
 }
 
+// This function fetches the appropriate instance property values for first level simple, structured property paths and
+// other levels of structured entity properties from the propertyPath.
+function getPropertyValuesFromInstance(currentProperty, entityInstance) {
+  let resultObject = {};
+  resultObject.propertyPath = currentProperty.propertyPath;
+
+  if(Array.isArray(entityInstance)) {
+    let propertyValuesArray = [];
+    entityInstance.forEach((instance) => {
+      let propertyValue = getPropertyPathValues(currentProperty, instance).propertyValue;
+      if(Array.isArray(propertyValue)) {
+        propertyValue.forEach((propertyValue) => {
+          propertyValuesArray.push(propertyValue);
+        });
+      } else {
+        propertyValuesArray.push(propertyValue);
+      }
+    });
+    resultObject.propertyValue = propertyValuesArray;
+  } else {
+    resultObject.propertyValue =  getPropertyPathValues(currentProperty, entityInstance).propertyValue;
+  }
+  return resultObject;
+}
+
+// This function fetches the appropriate instance for first level simple, structured property paths and other levels of
+// structured entity properties from the propertyPath.
+function getInstanceFromPropertyPath(currentProperty, entityInstance) {
```

Comment:
This naming is confusing - the only time we say "instance" is when referring to an "entity instance". Is this really a "value instance", or simply "property value"?